update(JS): web/javascript/reference/statements/for...of (#625)
webdoky/content: files/uk/web/javascript/reference/statements/for...of/index.md

{{jsSidebar("Statements")}}

**Інструкція `for...of`** виконує цикл, що обробляє послідовність значень, отриманих з [ітерованого об'єкта](/uk/docs/Web/JavaScript/Reference/Iteration_protocols#protokol-iteruvannia). До ітерованих об'єктів належать екземпляри вбудованих типів, як то {{jsxref("Array")}}, {{jsxref("String")}}, {{jsxref("TypedArray")}}, {{jsxref("Map")}}, {{jsxref("Set")}}, {{domxref("NodeList")}} (та інших колекцій DOM), а також об'єкт {{jsxref("Functions/arguments", "arguments")}}, породжені [генераторними функціями](/uk/docs/Web/JavaScript/Reference/Statements/function*) [генератори](/uk/docs/Web/JavaScript/Reference/Global_Objects/Generator) й визначені користувачем ітеровані об'єкти.

{{EmbedInteractiveExample("pages/js/statement-forof.html")}}

## Синтаксис

> ```js

Reviewer: ```suggestion
```js-nolint
```

> for (variable of iterable) statement;

Reviewer: ```suggestion
for (variable of iterable)
  statement;
```

Вихід з циклу `for...of` відбувається, коли завершується ітератор (коли метод ітератора `next()` повертає об'єкт, що містить `done: true`). Також для зміни звичайного ходу виконання можна використовувати інструкції контролю ходу виконання. [`break`](/uk/docs/Web/JavaScript/Reference/Statements/break) спричиняє вихід з циклу і перехід до першої інструкції після тіла циклу, натомість [`continue`](/uk/docs/Web/JavaScript/Reference/Statements/continue) призводить до пропуску решти інструкцій поточної ітерації та переходу до наступної ітерації.

Якщо з циклу `for...of` відбувається ранній вихід (наприклад, якщо зустрілася інструкція `break` чи була викинута помилка), то викликається метод [`return()`](/uk/docs/Web/JavaScript/Reference/Iteration_protocols#protokol-iteratora), щоб виконати прибирання.

> The `variable` part of `for...of` accepts anything that can come before the `=` operator. You can use {{jsxref("Statements/const", "const")}} to declare the variable as long as it's not reassigned within the loop body (it can change between iterations, because those are two separate variables). Otherwise, you can use {{jsxref("Statements/let", "let")}}.

Reviewer: ```suggestion
```

```js
function foo() {
  for (const value of arguments) {
    console.log(value);
  }
}
foo(1, 2, 3);
// 1
// 2
// 3
```

### Ітерування NodeList

Наступний приклад додає клас `read` до абзаців, що є безпосередніми нащадками елемента [`<article>`](/uk/docs/Web/HTML/Element/article), шляхом ітерування колекції DOM [`NodeList`](/uk/docs/Web/API/NodeList).

```js
const articleParagraphs = document.querySelectorAll('article > p');
for (const paragraph of articleParagraphs) {
  paragraph.classList.add('read');
}
```

### Ітерування ітерованого, визначеного користувачем

Ітерування об'єкта з методом `@@iterator`, що повертає самописний ітератор:

```js
const iterable = {
  [Symbol.iterator]() {
    let i = 1;
    return {
      next() {
        if (i <= 3) {
          return { value: i++, done: false };
        }
        return { value: undefined, done: true };
      },
    };
  },
};
for (const value of iterable) {
  console.log(value);
}
// 1
// 2
// 3
```

Ітерування об'єкта з генераторним методом `@@iterator`:

```js
const iterable = {
  *[Symbol.iterator]() {
    yield 1;
    yield 2;
    yield 3;
  },
};
for (const value of iterable) {
  console.log(value);
}
// 1
// 2
// 3
```

> _Ітеровані ітератори_ (ітератори з методом `[@@iterator]()`, що повертає `this`) є доволі поширеним підходом, щоб зробити ітератори застосовними в синтаксичних конструкціях, що очікують ітерованих, як то `for...of`.

Reviewer: Таке формулювання виглядає дивним, я б додавав "що саме є ітерованим". З одного боку - хочеться трішки натягувати сову на глобус, і розширювати українську мову, а з іншого - трохи бентежить, що результат іноді виглядає як переклад гуглотранслейтом зразка 16го року
```suggestion
_Ітеровані ітератори_ (ітератори з методом `[@@iterator]()`, що повертає `this`) є доволі поширеним підходом, щоб зробити ітератори застосовними в синтаксичних конструкціях, що очікують на ітеровані об'єкти, як то `for...of`.
```

Інструкція {{jsxref("Statements/for...in", "for...in")}} ітерує [перелічувані рядкові властивості](/uk/docs/Web/JavaScript/Enumerability_and_ownership_of_properties) об'єкта, натомість інструкція `for...of` ітерує значення, котрі [ітерований об'єкт](/uk/docs/Web/JavaScript/Reference/Iteration_protocols#protokol-iterovanoho) визначає для ітерування.

Наступний приклад демонструє різницю між циклом `for...of` та циклом `for...in` при застосуванні на {{jsxref("Array")}}.

```js
Object.prototype.objCustom = function () {};
Array.prototype.arrCustom = function () {};
const iterable = [3, 5, 7];
iterable.foo = 'агов';
for (const i in iterable) {
  console.log(i); // виведе "0", "1", "2", "foo", "arrCustom", "objCustom"
}
for (const i in iterable) {
  if (Object.hasOwn(iterable, i)) {
    console.log(i); // виведе "0", "1", "2", "foo"
  }
}
for (const i of iterable) {
  console.log(i); // виведе 3, 5, 7
}
```

Об'єкт `iterable` успадковує властивості `objCustom` та `arrCustom`, адже має у своєму [ланцюжку прототипів](/uk/docs/Web/JavaScript/Inheritance_and_the_prototype_chain) і `Object.prototype`, і `Array.prototype`.

> The `for...in` loop logs only [enumerable properties](/uk/docs/Web/JavaScript/Enumerability_and_ownership_of_properties) of the `iterable` object. It doesn't log array _elements_ `3`, `5`, `7` or `"hello"` because those are not _properties_ — they are _values_. It logs array _indexes_ as well as `arrCustom` and `objCustom`, which are actual properties. If you're not sure why these properties are iterated over, there's a more thorough explanation of how [array iteration and `for...in`](/uk/docs/Web/JavaScript/Reference/Statements/for...in#array_iteration_and_for...in) work.

Reviewer: ```suggestion
```

> Цикл `for...in` виводить лише [перелічувані властивості](/uk/docs/Web/JavaScript/Enumerability_and_ownership_of_properties) об'єкта `iterable`. Він не виводить _елементи_ масиву, `3`, `5`, `7` чи `"агов"`, бо вони не є _властивостями_: вони є _значеннями_. Він виводить _індекси_ масиву, так само як `arrCustom` і `objCustom`, що є справжніми властивостями.

Reviewer: ```suggestion
Цикл `for...in` виводить лише [перелічувані властивості](/uk/docs/Web/JavaScript/Enumerability_and_ownership_of_properties) об'єкта `iterable`. Він не виводить _елементи_ масиву, `3`, `5`, `7` чи `"агов"`, бо вони не є _властивостями_: вони є _значеннями_. Він виводить _індекси_ масиву, так само як `arrCustom` і `objCustom`, що є справжніми властивостями. Якщо немає певності в тому, чому ітеруються саме ці властивості — є більш поглиблене роз'яснення, як працюють [ітерування масиву та цикл `for...in`](/uk/docs/Web/JavaScript/Reference/Statements/for...in#array_iteration_and_for...in).
```

Author: ```suggestion
Цикл `for...in` виводить лише [перелічувані властивості](/uk/docs/Web/JavaScript/Enumerability_and_ownership_of_properties) об'єкта `iterable`. Він не виводить _елементи_ масиву, `3`, `5`, `7` чи `"агов"`, бо вони не є _властивостями_: вони є _значеннями_. Він виводить _індекси_ масиву, так само як `arrCustom` і `objCustom`, що є справжніми властивостями. Якщо немає певності в тому, чому ітеруються саме ці властивості — є більш поглиблене роз'яснення, як працюють [ітерування масиву та цикл `for...in`](/uk/docs/Web/JavaScript/Reference/Statements/for...in#iteruvannia-masyvu-i-tsykl-forin).
```

Другий цикл – подібний до першого, але він використовує {{jsxref("Object.hasOwn()")}}, аби перевіряти, чи є знайдена перелічувана властивість власною властивістю об'єкта, тобто не успадкованою. Якщо це так, то властивість виводиться. Властивості `0`, `1`, `2` і `foo` – виводяться, тому що є власними. Властивості `arrCustom` і `objCustom` не виводяться, бо є успадкованими.

> Цикл `for...of` ітерує й виводить _значення_, котрі `iterable`, як масив (а масиви є [ітерованими](/uk/docs/Web/JavaScript/Reference/Global_Objects/Array/@@iterator)), визначає для ітерування. Демонструються _елементи_ об'єкта – `3`, `5`, `7`, але жодна з _властивостей_ об'єкта.

Reviewer: довго думав, до чого тут "визначає для ітерування", якщо достатньо просто "значення [є] `iterable`, як масив"... Пропоную додати акценту, аби не виникало питань, хто то визначає
```suggestion
Цикл `for...of` ітерує й виводить _значення_, котрі об'єкт `iterable`, як масив (а масиви є [ітерованими](/uk/docs/Web/JavaScript/Reference/Global_Objects/Array/@@iterator)), визначає для ітерування. Демонструються _елементи_ об'єкта – `3`, `5`, `7`, але жодна з _властивостей_ об'єкта.
```
